File: SimpleEngine/SimpleWorld.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

namespace SimpleEngine
{
    public class SimpleWorld : World
    {
        private List<Renderable> worldRenderables = new List<Renderable>();
        private List<Renderable> cameraRenderables = new List<Renderable>();
        private List<Entity> entities = new List<Entity>();

        delegate void RenderableAddTarget(Renderable r);

        public SimpleWorld(MediaLoader loader, string file)
        {
            using (var s = loader.FS.open(file))
            {
                XmlElement level = Xml.Open(Xml.FromStream(s), "level");
                addMeshes(loader, level["world"], add);
                addMeshes(loader, level["camera"], addCamera);
                addEntities(loader, level["entity"]);
            }
        }

        private void addEntities(MediaLoader loader, XmlElement level)
        {
            foreach (XmlElement entity in Xml.ElementsNamed(level, "entity"))
            {
                string t = Xml.GetAttributeString(entity, "type");
                if (t[0] == '_') continue; // ignore for now
                string name = Xml.GetAttributeString(entity, "name");
                string meshpath = t + ".gob";
                vec3 pos = GetPosition(entity["position"]);
                quat rot = GetRotation(entity["rotation"]);

                using (var gobstream = loader.FS.open(meshpath))
                {
                    addEntity(Entity.Create(loader, name, Xml.Open(Xml.FromStream(gobstream), "object"), pos, rot));
                }
            }
        }

        private void addMeshes(MediaLoader loader, XmlElement level, RenderableAddTarget target)
        {
            foreach (XmlElement entity in Xml.ElementsNamed(level, "entity"))
            {
                string t = Xml.GetAttributeString(entity, "type");
                string name = Xml.GetAttributeString(entity, "name");
                string meshpath = t + ".mdf";
                MeshInstance mesh = new MeshInstance(loader.fetch<Mesh>(meshpath));
                mesh.pos = GetPosition(entity["position"]);
                mesh.rot = GetRotation(entity["rotation"]);
                target(mesh);
            }
        }

        private static vec3 GetPosition(XmlElement e)
        {
            return new vec3(dp(e, "x"), dp(e,"y"), dp(e, "z"));
        }

        private static float dp(XmlElement e, string p)
        {
            string val = Xml.GetAttributeString(e, p);
            return float.Parse(val.Replace('.', ','));
        }

        private static quat GetRotation(XmlElement e)
        {
            return new quat(dp(e, "w"), new vec3(dp(e, "x"), dp(e, "y"), dp(e, "z")));
        }

        public override void add(Renderable r)
        {
            worldRenderables.Add(r);
        }

        public override void remove(Renderable r)
        {
            worldRenderables.Remove(r);
        }

        public override void worldSendTo(RenderList list)
        {
            sendToList(list, worldRenderables);
        }

        private static void sendToList(RenderList target, List<Renderable> container)
        {
            foreach (Renderable r in container)
            {
                r.sendToRenderer(target);
            }
        }

        public override void addCamera(Renderable r)
        {
            cameraRenderables.Add(r);
        }

        public override void cameraSendTo(RenderList list)
        {
            sendToList(list, cameraRenderables);
        }

        public override void addEntity(Entity ent)
        {
            entities.Add(ent);
            foreach (Renderable r in ent.visual)
            {
                add(r);
            }
        }
    }
}

File: src/engine/simpleworld.cc
#include "engine/simpleworld.h"

#include <functional>
#include <memory>
#include <string>
#include <vector>

#include "engine/angle.h"
#include "engine/cpp.h"
#include "engine/filesystem.h"
#include "engine/math.h"
#include "engine/medialoader.h"
#include "engine/mesh.h"
#include "engine/meshinstance.h"
#include "engine/opengl.h"
#include "engine/quat.h"
#include "engine/vec3.h"
#include "engine/world.h"

namespace simple
{
    namespace
    {
        float ReadFloat(std::shared_ptr<xml::Element> element, const std::string& attribute_name)
        {
            const std::string value = xml::GetAttributeString(element, attribute_name);
            return math::ParseFloat(value);
        }

        Angle ReadAngle(std::shared_ptr<xml::Element> element, const std::string& attribute_name)
        {
            const auto f = ReadFloat(element, attribute_name);
            return Angle::FromDegrees(f);
        }

        vec3 GetPosition(std::shared_ptr<xml::Element> e)
        {
            return vec3(ReadFloat(e, "x"), ReadFloat(e, "y"), ReadFloat(e, "z"));
        }

        quat GetRotation(std::shared_ptr<xml::Element> e)
        {
            return quat::FromYawPitchRoll(ReadAngle(e, "y"), ReadAngle(e, "p"), ReadAngle(e, "r"));
        }

        void SendToList(RenderList* target, const std::vector<std::shared_ptr<Renderable>>& container)
        {
            for (const auto& renderable : container)
            {
                renderable->SendToRenderer(target);
            }
        }
    }

    SimpleWorld::SimpleWorld(MediaLoader* loader, const std::string& file)
    {
        auto absolute_path = loader->FS()->Open(file);
        auto level = xml::Open(absolute_path, "level");
        AddMeshes(loader, level->GetChild("world"), [this](auto x) { this->Add(x); });
        AddMeshes(loader, level->GetChild("camera"), [this](auto x) { this->AddCameraRenderable(x); });
        AddEntities(loader, level->GetChild("entity"));
    }

    void SimpleWorld::AddEntities(MediaLoader* loader, std::shared_ptr<xml::Element> level)
    {
        for (const std::shared_ptr<xml::Element>& entity : xml::ElementsNamed(level, "entity"))
        {
            auto type = xml::GetAttributeString(entity, "type");
            if (type[0] == '_')
                continue;  // ignore for now
            std::string name = xml::GetAttributeString(entity, "name");
            std::string meshpath = type + ".gob";
            vec3 pos = GetPosition(entity->GetChild("position"));
            quat rot = GetRotation(entity->GetChild("rotation"));

            auto gobstream = loader->FS()->Open(meshpath);
            AddEntity(Entity::Create(loader, name, xml::Open(gobstream, "object"), pos, rot));
        }
    }

    void SimpleWorld::AddMeshes(MediaLoader* loader, std::shared_ptr<xml::Element> level, const RenderableAddTarget& target)
    {
        for (std::shared_ptr<xml::Element> entity : xml::ElementsNamed(level, "mesh"))
        {
            // auto name = Xml::GetAttributeString(entity, "name");
            auto meshpath = xml::GetAttributeString(entity, "file");
            // auto meshpath = t + ".mdf";
            auto mesh = std::make_shared<MeshInstance>(loader->GetOrLoad<Mesh>(meshpath));
            mesh->position = GetPosition(entity->GetChild("position"));
            mesh->rotation = GetRotation(entity->GetChild("rotation"));
            target(mesh);
        }
    }

    void SimpleWorld::Add(std::shared_ptr<Renderable> r)
    {
        world_renderables.push_back(r);
    }

    void SimpleWorld::Remove(std::shared_ptr<Renderable> renderable)
    {
        Erase(world_renderables, [renderable](std::shared_ptr<Renderable>& other) -> bool { return other == renderable; });
    }

    void SimpleWorld::SendWorldRenderablesToList(RenderList* list)
    {
        SendToList(list, world_renderables);
    }

    void SimpleWorld::AddCameraRenderable(std::shared_ptr<Renderable> renderable)
    {
        camera_renderables.push_back(renderable);
    }

    void SimpleWorld::SendCameraRenderablesToList(RenderList* list)
    {
        SendToList(list, camera_renderables);
    }

    void SimpleWorld::AddEntity(std::shared_ptr<Entity> entity)
    {
        entities.push_back(entity);
        for (auto r : entity->renderables)
        {
            Add(r);
        }
    }

    void SimpleWorld::ClearScreen()
    {
        glClear(GL_COLOR_BUFFER_BIT);
    }
}
